ckan/lib/search/index.py

```diff
@@ -279,6 +279,13 @@ def index_package(self, pkg_dict, defer_commit=False):
 
         assert pkg_dict, 'Plugin must return non empty package dict on index'
 
+        # permission labels determine visibility in search, can't be set
+        # in original dataset or before_index plugins
+        labels = lib_plugins.get_permission_labels()
+        dataset = model.Package.get(pkg_dict['id'])
+        pkg_dict['permission_labels'] = labels.get_dataset_labels(
+            dataset) if dataset else [] # TestPackageSearchIndex-workaround
```

Comment:
I started building this interface expecting user and dataset dicts, but I ended up switching to the models instead.

These permissions are partially enforced below the logic layer (lib.search). Calling back up to the logic layer to get the dict version of the user object doesn't make sense, also it would make what used to be a super-cheap package_show auth function way more expensive to call. It's also inconsistent to have one part of the interface take a model and the other part take a dict.

Also with @TkTech's plugin extras changes we might have extras attached to users and datasets that aren't part of the dict presented to the user (e.g. permissions or labels applied by someone else) and having the model will make those extras easier to access.